### ISSUE TITLE
npm login _authToken configuration support missing in npm endpoint when npm registry and collection are on the same domain

### ISSUE THREAD

**Issue by paulgdunn** [microsoft/azure-pipelines-tasks#7316]
## Environment
- Server - TFS on-premises - TFS 2018 U2 and TFS 2017 U3
    
- Agent - Private: 
    - Windows Server 2012 R2

## Issue Description

1) the password field cannot support longer length'ed  passwords in order to support _authToken (npm login)

2) npm login / _authToken not supported

npm build task and npm endpoint does not support _authToken configuration.  For example, using the command "npm login" and provide credentials for a npm registry/repository, an entry will be added to the .npmrc that has _authToken="a token value".  It would be useful to provide the ability to use _authToken (npm login) within the npm endpoint.

As a workaround, on our test TFS 2017 server that matches our current production server version, we found a solution by building a customized extension in TFS.  We were able to reverse engineer the npm endpoint definition.  We uploaded our version of the npm endpoint definition within a custom TFS Extension.  The extension definition only contains our custom npm endpoint.  We found that we were able to get the _authToken to work with our private locally hosted NPM Artifactory repository by using the value of the _authToken as the value of the password.  This was through using the username/password configuration (not the PATs).  However, we also had to increase the password field length to "maxLength":1000.  We also uploaded the latest version of the npm build task from github.

On our test TFS 2018, we cannot get the customized version of the npm endpoint definition with the maxLength:1000 to take effect.  Rather, the length seems to be out-of-box length of 300.

In conclusion, 
It would be best if the npm endpoint definition could support a scenario of the "npm login" where the _authToken="value" could be utilized.   plus, max Length of token of 1000 (or even larger)

Alternatively, we could use a longer maxLength:1000 (or even longer length just in case) for the password npm endpoint definition in TFS 2018, which would likely work for our npm registry/repository.

**Comment by jotaylo** [comment]
Does this PR address your issues? #6440
If you swap back to specifying a token (which has a 2000 character limit) it should now be supplied as _authToken

**Comment by paulgdunn** [comment]
After having a second look at the linked item #6440 and the commit from Feb 21, it seems that the condition and block starting with line 51: https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/Common/npm-common/npmregistry.ts#L51 is not compatible with anyone that would have an NPM registry and TFS server deployed on-premise where both servers are on the same domain.  And where the NPM registry would use the form of Token authentication with _authToken entry in the .npmrc,

So, that I was able to edit the following to get our publish to work: 

from this:
```
            // To the reader, this could be optimized here but it is broken out for readability
            if (endpointAuth.scheme === 'Token'){
                if (collectionUrlDomain === epUrlDomain){
                    // Same domain _OR_ matching IP addrs.  Use PAT+Basic
                    isVstsTokenAuth = true;
                } else if (epUrlDomain.toUpperCase() === 'VISUALSTUDIO.COM'){
                    // If the endpoint is VSTS, full stop.  Use PAT+Basic
                    isVstsTokenAuth = true;
                }
            }
```

to this:

```
            // To the reader, this could be optimized here but it is broken out for readability
            if (endpointAuth.scheme === 'Token') {
                if (collectionUrlDomain === epUrlDomain) {
                    // Same domain _OR_ matching IP addrs.  Use PAT+Basic
                    isVstsTokenAuth = false;
                }
                else if (epUrlDomain.toUpperCase() === 'VISUALSTUDIO.COM') {
                    // If the endpoint is VSTS, full stop.  Use PAT+Basic
                    isVstsTokenAuth = true;
                }
            }

```
When 
`isVstsTokenAuth = false; `
then, the "Token" ends up being the desired "_authToken" entry in the .npmrc and not the VSTS type.

```
            case 'Token':
                let apitoken = endpointAuth.parameters['apitoken'];
                if (!isVstsTokenAuth){
                    // Use Bearer auth as it was intended.
                    auth = nerfed + ":_authToken=" + apitoken + lineEnd;
                }else{
                    // VSTS does not support PATs+Bearer only JWTs+Bearer


```
Can you update the implementation to account for our needs?  I'm not sure what the purpose of lines 51 through 53 are meant to do?  Our needs are that we want to be able to have:
```
                    // Use Bearer auth as it was intended.
                    auth = nerfed + ":_authToken=" + apitoken + lineEnd;
```
 in the same manner as using "npm login".  Plus, our NPM registry and TFS on-premise deployment are hosted on the same domain.

**Comment by paulgdunn** [comment]
In my opinion, it would be best if the implementation of the npm endpoint/service definition was the same as the nuget endpoint/service definition.   NPM endpoints only have Basic and PATS.  NuGet endpoints have Basic, PATS, and Token.  If the endpoint definition for npm had 3 options, then the task could be implemented to support Basic, VSTS PATS, and _authToken.  Also, it would be best if the endpoints could be defined adjacent to the Task implementation and checked into github.

**Comment by elbatk** [comment]
Hi @paulgdunn , sorry for the long delay here. We are taking a hard look at our tasks in Azure Pipelines this quarter and we're looking to transition into lighter-weight authentication tasks. I'm copying your feedback in our feature tracking to make sure it's considered there.

**Comment by aasim** [comment]
Hi - We have recently released the NpmAuthenticate task that can be used in a pipeline to authenticate with Azure Artifacts and other Npm repositories. The design philosophy behind this task is that we would authenticate all the repositories you describe in this task, and get out of the way leaving you fully in control of running your npm commands in the pipeline as you see fit.

You can read more about the NpmAuthenticate task here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/package/Npm-authenticate?view=azure-devops. It also gives some examples on how to configure the task.

I think this task is a good fit for your use case, you can use the NpmAuthenticate task to perform the authentication, and then use the same npm command you use locally to run in the pipeline. Let me know if you need any more help in setting this up.

Aasim
